### ISSUE TITLE
Maximum call stack size exceeded at BinarySearchTree.compareThings

### ISSUE THREAD

**Issue by vladmandic** [seald/nedb#1]
On database ~500MB with 10K records in a single table, there are frequent errors:

```
/home/vlado/dev/pigallery/node_modules/.pnpm/@seald-io+nedb@2.0.3/node_modules/@seald-io/nedb/lib/model.js:175
const compareThings = (a, b, _compareStrings) => {
                      ^

RangeError: Maximum call stack size exceeded
    at BinarySearchTree.compareThings [as compareKeys] (/home/vlado/dev/pigallery/node_modules/.pnpm/@seald-io+nedb@2.0.3/node_modules/@seald-io/nedb/lib/model.js:175:23)
    at BinarySearchTree.delete (/home/vlado/dev/pigallery/node_modules/.pnpm/@seald-io+binary-search-tree@1.0.2/node_modules/@seald-io/binary-search-tree/lib/bst.js:348:14)
    at BinarySearchTree.delete (/home/vlado/dev/pigallery/node_modules/.pnpm/@seald-io+binary-search-tree@1.0.2/node_modules/@seald-io/binary-search-tree/lib/bst.js:354:40)
    at BinarySearchTree.delete (/home/vlado/dev/pigallery/node_modules/.pnpm/@seald-io+binary-search-tree@1.0.2/node_modules/@seald-io/binary-search-tree/lib/bst.js:354:40)
    at BinarySearchTree.delete (/home/vlado/dev/pigallery/node_modules/.pnpm/@seald-io+binary-search-tree@1.0.2/node_modules/@seald-io/binary-search-tree/lib/bst.js:354:40)
```

This most commonly happens during calls to `delete` function, but also happens during any number of other operation.

Example that causes error - record itself is perfectly valid and can be read without issues, just cannot be deleted
note this happens to only small number of records in database, but its fully reproducible with specific record in question

```js
db.findOne({ image: param }, (err, res) => {
  if (res) db.remove({ image: data.image }, { multi: false }, (err, count) => {
    console.log(err, count);
  });
});
```

Workaround is to run `node --stack-size=8192` and then there are no issues,
But this really should not be necessary for a DB of this size - I'm guessing recursive calls for large number of cells need some rethinking, perhaps using something like `setTimeout(() => { next(); });` to give node a chance to clear a stack before diving into next recursion.

Environment: NodeJS 16.0.0 on Ubuntu 21.04, `@seald-io/nedb` 2.0.3

Crosslinking issue with <https://github.com/jrop/nedb-promise/issues/14>

**Comment by eliot-akira** [comment]
I was able to reproduce this issue with the following test.

```js
const Datastore = require('@seald-io/nedb')

const db = new Datastore({ filename: 'test.db', autoload: true })
const maxItems = 10000

;(async () => {

  for (let i=0; i < maxItems; i++) {
    await new Promise((resolve, reject) => {
      db.insert({ i }, function (err, doc) {
        err ? reject(err) : resolve()
      })
    })
    console.log('Created', i)
  }

  for (let i=maxItems - 1; i >= 0; i--) {
    await new Promise((resolve, reject) => {
      db.remove({ i }, function (err) {
        err ? reject(err) : resolve()
      })
    })
    console.log('Removed', i)
  }

})()
```

Running the above with the following command:

```sh
node --stack-size=1000 index.js
```

..Results in the same error:

```
RangeError: Maximum call stack size exceeded
    at BinarySearchTree.compareThings (node_modules/@seald-io/nedb/lib/model.js:175:23)
```

I also found a possible solution. The first line of `@seald-io/nedb/lib/indexes.js` is currently:

```js
const BinarySearchTree = require('@seald-io/binary-search-tree').BinarySearchTree
```

..Changing it to use AVLTree, which the original binary-search-tree implementation used:

```js
const BinarySearchTree = require('@seald-io/binary-search-tree').AVLTree
```

Then the test case no longer throws.

**Comment by tex0l** [comment]
Hi !
I don't know why I haven't received a notification for this issue, sorry for the delay.

I've seen the PR, I don't know why I switched to a BinarySearchTree instead of an AVLTree when refactoring, this looks very much like a mistake on my end.

I'm running some tests and will fix this in an upcoming patch.

**Comment by tex0l** [comment]
Fixed in https://github.com/seald/nedb/commit/f68464d5c4e9aee4fc59ea4d8df42df20a26ba08

Released in 2.0.4
